src/helm/benchmark/scenarios/grammar.py

```diff
@@ -0,0 +1,148 @@
+from collections import defaultdict
+from dataclasses import dataclass, field, replace
+from typing import List, Optional
+from helm.common.hierarchical_logger import hlog
+
+import dacite
+import re
+import yaml
+
+
+ROOT_CATEGORY = "Root"
+
+
+@dataclass
+class Expansion:
+    """
+    Example: text = "this is a ${Topic}"
+    """
+
+    text: str
+    tags: List[str] = field(default_factory=list)
+
+    def __post_init__(self):
+        self.children = re.split(r"(\${\w+})", self.text)
+        self.categories = [x for x in [get_category(child) for child in self.children] if x is not None]
+
+
+@dataclass(frozen=True)
+class GrammarRule:
+    """
+    Represents a set of grammar rules:
+
+            <category> -> expansions[0]
+            <category> -> expansions[1]
+            ...
+    """
+
+    category: str
+    expansions: List[Expansion]
+    tags: List[str] = field(default_factory=list)
+
+
+def get_category(text: str) -> Optional[str]:
+    """
+    Example: "${Root}" => "Root"
+    """
+    if text.startswith("${") and text.endswith("}"):
+        return text[2:-1]
+    return None
+
+
+@dataclass
+class Grammar:
+    rules: List[GrammarRule]
+
+    def __post_init__(self):
```

Comment:
Philosophically, I would prefer to treat all dataclasses as immutable tuples (i.e. always set `frozen=True`). I think `__post_init__` hurts the usability of dataclasses because now you have to reason about fields settable at init time vs fields settable after init time vs fields that are technically computed values and shouldn't be manually set.

In particular, I think `category_to_rules` should be a [`cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property) rather than a `field` (because it's a function of `rules`).

The same should apply to `Expansion` except there's one point later where you modify `children` without modifying `text`... so it's not a function?

Reply:
Good idea, switched to cached_property.